### PR TITLE
Optimizes INVOKE_ASYNC by making it a macro and avoiding a proc call

### PR DESCRIPTION
**Pull request by Kapu1178**
## About The Pull Request
This is quite literally the same behavior but faster, and also catches improper arguments better than the old macro/proc approach.

Credit to Lohikar for writing the macro.
Port of https://github.com/DaedalusDock/daedalusdock/pull/196

Also, `world.ImmediateInvokeAsync()` never set a return value, so expecting one was never valid behavior.

At MSO's request, the documentation of `spawn(-1)`:

As per the reference, calling `spawn()` with a negative value will execute the spawned code until a blocking action (such as `sleep()`) is encountered. Then, it will step outside of the spawned code, and continue the proc. This is the same behavior as calling a `waitfor = FALSE` proc. Specifically, under the hood, `spawn(-1)` creates a copy of the callstack like `sleep()`, incase the spawned code is blocked and needs to be rescheduled.

As an added bonus, `spawn(-1)` silences SHOULD_NOT_SLEEP errors, whereas `waitfor = FALSE` does not.

## Why It's Good For The Game

ITS FREE FUCKING CPU TIME

### CODE REVIEW

File: code/__DEFINES/callbacks.dm
Comment (Mothblocks) [inline code/__DEFINES/callbacks.dm:4]: All these names need to be fully written out and snake_case

Also preferably this should be multilined because it's really hard to read

File: code/datums/components/payment.dm
Comment (Mothblocks) [inline code/datums/components/payment.dm:126]:
```suggestion
				successfully_put_in_hands = paying_customer.put_in_hands(holochange)
```

File: code/datums/components/construction.dm
Comment (Mothblocks) [inline code/datums/components/construction.dm:36]: I feel like I would rather we have a proc that actually does give a return value because this is only going to serve to confuse everyone who has never seen this construct before (including maintainers) and potentially lead it to be used in places it doesn't belong

File: code/datums/components/construction.dm
Comment (Mothblocks) [inline code/datums/components/construction.dm:37]: `check_step` should probably `set waitfor = FALSE` instead, then we can `return check_step(I, user)` directly

**Review by Mothblocks — CHANGES_REQUESTED** [review]

**Comment by Mothblocks** [comment]
Also looking at random rounds, ImmediateInvokeAsync has a self cost of like 2 milliseconds????

**Comment by Mothblocks** [comment]
Per contribution guidelines:

> Should you be optimizing a routine you must provide proof by way of profiling that your changes are faster.

This seems like a readability loss for effectively nil gain since we don't use this anywhere hot from what I can tell

**Comment by Mothblocks** [comment]
image

File: code/datums/components/construction.dm
Comment (Kapu1178) [reply code/datums/components/construction.dm:37]: SHOULD_NOT_SLEEP() is a weird lint. It is EXTREMELY aggressive and ignores `set waitfor = FALSE`, it would require an UNLINT() wrap which is uglier imo.

**Comment by Kapu1178** [comment]
> Per contribution guidelines:
> 
> > Should you be optimizing a routine you must provide proof by way of profiling that your changes are faster.
> 
> This seems like a readability loss for effectively nil gain since we don't use this anywhere hot from what I can tell

It removes potential to expect a return value from INVOKE_ASYNC (because you cant assign a var to it, the compiler will cry), alongside catching other misuses of INVOKE_ASYNC. It caught `INVOKE_ASYNC(emote("scream"))` on my codebase already.

**Comment by Kapu1178** [comment]
> Please do not add new uses of spawn. I hate spawn

spawn fear is extremely outdated. Byond's scheduler is alot smarter than it used to be, and its the exact same thing as a `set waitfor = FALSE` proc without the implicit nature of a proc being pragma'd to be async.

File: code/datums/components/construction.dm
Comment (Kapu1178) [reply code/datums/components/construction.dm:36]: check_step() has a return value. It sleeps at some point in it's chain, which is why its in a spawn and its return value is treated as valid.

**Comment by LemonInTheDark** [comment]
It confuses people, and I don't like the code that using it produces.
Byond scheduler still sucks, but it's unrelated to my dislike of the tool.

**Comment by Kapu1178** [comment]
> Per contribution guidelines:
> 
> > Should you be optimizing a routine you must provide proof by way of profiling that your changes are faster.
> 
> This seems like a readability loss for effectively nil gain since we don't use this anywhere hot from what I can tell

Profiling the actual impact in prod is kind of impossible, but benchmarks show that its consistently about 11% faster, give or take a few %.

Code used for profiling:
```
#define GLOBAL_PROC "some magic bullshit"
#define INVOKE_ASYNC(TRG, PR, ARG...) if ((TRG) == GLOBAL_PROC) { spawn (-1) call(PR)(ARG); } else { spawn(-1) call(TRG, PR)(ARG); }

/world/proc/ImmediateInvokeAsync(thingtocall, proctocall, ...)
  set waitfor = FALSE

  if (!thingtocall)
    return

  var/list/calling_arguments = length(args) > 2 ? args.Copy(3) : null

  if (thingtocall == GLOBAL_PROC)
    call(proctocall)(arglist(calling_arguments))
  else
    call(thingtocall, proctocall)(arglist(calling_arguments))

var/global/a = 1
var/global/b = 2
/proc/testproc()
  return global.a + global.b

MAIN
  TICKBENCH("spawn", INVOKE_ASYNC(GLOBAL_PROC, /proc/testproc))
  TICKBENCH("invoke async", world.ImmediateInvokeAsync(GLOBAL_PROC, /proc/testproc))
  TICKBENCH("spawn again", INVOKE_ASYNC(GLOBAL_PROC, /proc/testproc))
  TICKBENCH("invoke async again", world.ImmediateInvokeAsync(GLOBAL_PROC, /proc/testproc))
  ```

**Comment by Kapu1178** [comment]
> It confuses people, and I don't like the code that using it produces. Byond scheduler still sucks, but it's unrelated to my dislike of the tool.

Abstracting such an important part of the engine isn't really the play imho. Hell, I'd even see the merit in something like
```
#define ASYNC spawn(-1)
```
simply so people dont have to question why -1 does what it does

**Comment by Kapu1178** [comment]
that isnt the button i pressed. github sucks.

**Comment by Mothblocks** [comment]
So we are saving 0.2ms

File: code/datums/components/construction.dm
Comment (Mothblocks) [reply code/datums/components/construction.dm:36]: I'm talking about `spawn(-1)` itself

Comment (Mothblocks) [reply code/datums/components/construction.dm:36]: This is why I would much rather `ImmediateInvokeAsync` give the return value instead

**Comment by Mothblocks** [comment]
> It removes potential to expect a return value from INVOKE_ASYNC (because you cant assign a var to it, the compiler will cry), alongside catching other misuses of INVOKE_ASYNC. 

I think the semantics of `value = INVOKE_ASYNC()` make as much sense as creating a spawn and taking the return value from that

**Comment by Mothblocks** [comment]
Nevermind the numbers I'm reading are in seconds, I didn't expect that, carry on :+1:

**Comment by Kapu1178** [comment]
> So we are saving 0.2ms

0.2 ms, catches more errors at compile time, and if someone who knows how to write regex wanted to, could completely remove the need for `PROC_REF()` with a small tweak to the macro itself. I hate PROC_REF.

The new syntax could easily be `INVOKE_ASYNC(user.someproc(my_arg))`, which is WAY nicer than the current syntax.

File: code/modules/vehicles/mecha/mecha_construction_paths.dm
Comment (LemonInTheDark) [inline code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: I do not like using this, straight up. spawn shouldn't be in "user space"

Comment (Kapu1178) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: We split the difference and make an `ASYNC` macro thats just `spawn(-1)`. Deal?

Comment (LemonInTheDark) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: the code should just be less stupid

Comment (Kapu1178) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: i agree but out of scope

File: code/datums/components/construction.dm
Comment (Kapu1178) [reply code/datums/components/construction.dm:36]: imo its more hassle to explain "Using this can return a return value but it also wont under these conditions" than to just educate people on how async works as a whole at this point. At what point are you abstracting so much that it's easier to just explain the core concept than to try to lock it away.

File: code/modules/vehicles/mecha/mecha_construction_paths.dm
Comment (LemonInTheDark) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: in scope spawn shouldn't be in user space

Comment (Kapu1178) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: its either spawn(), a couple UNLINT(), or PR close because im not refactoring mech code for this.

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]
~~Not allowed. Do not merge.~~ i was wrong but also kapu didn't document a thing. Please see my newer review for changes requested.

**Comment by MrStonedOne** [comment]
`set waitfor = FALSE` is not the same as spawn.

The primary and most common pattern this is for is

```dm
for(var/blah in listofblahs)
	blah.proc_that_sleeps_but_i_dont_want_to_have_to_wait_for_the_sleep_to_finish_to_process_the_next_item()
```

in this pattern `spawn` has to copy the proc/proc vars from the stack to a forked proc with an implied return and then enter the timer queue, wake up next tick, run the thing, then when the thing you run wants to sleep, this has to happen again.

with `spawn`, each item in the list takes 2 trips thru the byond sleeping proc queue.

however, with `set waitfor = FALSE`, it does nothing, and only has to enter the timer queue when it gets to the actual sleep/blocking operation, meaning only 1 trip thru the sleeping procs queue

Entering this queue is not free, and all things that enter that tick to run next tick has to step over all other things that entered that tick as byond ensures enter order -> fire order, so a for loop that calls things that sleeps will quickly balloon in overhead when using spawn.

`set waitfor = FALSE` was created to massively speed up explosion code and this pr would undo a lot of that.

File: code/datums/components/construction.dm
Comment (MrStonedOne) [reply code/datums/components/construction.dm:37]: > and ignores set waitfor = FALSE

I required they do this because set waitfor = FALSE is often used on should not sleep procs as a backup. should not sleep is a compile time lint of a runtime typed language, there are codepaths that can come up at runtime that do not trigger runtime errors that can cause different procs to run then the compile time code might have been expecting. So any proc called by the mc has to have set waitfor = FALSE, regardless of should not sleep status, just incase.

**Comment by MrStonedOne** [comment]
To profile something like that, where the cost is spread between fires, you have to do it by hand, on your computer, no bots allowed.

You must profile by cpu time in task manager or process explorer on virgin runs (new dd process between runs, no start/stop bullshit)

This is the ONLY way to totally control for all variables when talking about profiling something that crosses ticks.

You have to profile:

* The first sleep enter (from the spawn) (done) 
* The first sleep exit (from the spawn) (not profiled in your profile)
* The first proc call overhead (set waitfor = true only)
* The time spent running the main proc (this is where you made the biggest mistake, BENCH would include the proc call overhead cost of running both procs in the waitfor tests **but not the spawn tests** which wouldn't even the one proc so you are just seeing cost that exists in all runs, but only seeing the cost reflected in the output of one run, making you think it was slower then it was)
* The second sleep enter (from the sleep within the proc) (only profiled on set waitfor = TRUE)


it should be such that world/New() starts the profile and prints to world.log when it is done. then you look at the cpu time colume in process explorer or task manager to see exactly how much processor time byond spent doing that test.

This ensures your profile includes all overhead, including cross tick overhead from modifying the sleeping proc queue twice as often.

**Comment by Kapu1178** [comment]
> **`set waitfor = FALSE` is not the same as `spawn`.**
> 
> The primary and most common pattern this is for is
> 
> ```dm
> for(var/blah in listofblahs)
> 	blah.proc_that_sleeps_but_i_dont_want_to_have_to_wait_for_the_sleep_to_finish_to_process_the_next_item()
> ```
> 
> in this pattern `spawn` has to copy the proc/proc vars from the stack to a forked proc with an implied return and then enter the forked proc into the timer queue with a time of `0`, wake up next tick, run the thing, then when the thing you run wants to sleep, this has to happen again. **FOR EVERY ITEM IN THE LIST**
> 
> with `spawn`, each item in the list takes 2 trips thru the byond sleeping proc queue.
> 
> however, with `set waitfor = FALSE`, it does nothing, and only has to enter the timer queue when it gets to the actual sleep/blocking operation, meaning only 1 trip thru the sleeping procs queue
> 
> Entering this queue is not free, and all things that enter that tick to run next tick has to step over all other things that entered that tick as byond ensures enter order -> fire order, so a for loop that calls things that sleeps will quickly balloon in overhead when using spawn.
> 
> `set waitfor = FALSE` was dug out of undocumented hell to massively speed up explosion code and this pr would undo a lot of that.

You do realise that waitfor= FALSE procs do that too, right?

`sleep(-1)` is extremely special in that it explicitly does not sleep. It does not go through scheduler hell.
image

You are absolutely correct in that spawn(-1) copies the callstack even if the proc even if it doesn't sleep. In general, INVOKE_ASYNC() procs will sleep, that's why it exists in the first place. In the extremely, extremely, extremely few instances where that matters, it can be adjusted accordingly. 

During benchmarking to see if this is actually faster, I found that in most cases, `spawn(-1)` is faster than `world.ImmediateInvokeAsync()` by 11-ish percent.

File: code/datums/components/construction.dm
Comment (Kapu1178) [reply code/datums/components/construction.dm:37]: Fair play.

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]
You must comment **all** uses of spawn (-1) explaining in detail how it is different from spawn and does not do what spawn normally does.

i would suggest doing this by adding spawn (-1) to a macro called ASYNC and putting the comment on that macro. =P

You should also detail the -1 discovery in the pr body so this pr makes sense.

**Comment by Kapu1178** [comment]
Big dump of the discord convo
image
image
image
image
image
image
image
image
image
image

File: code/__DEFINES/callbacks.dm
Comment (MrStonedOne) [inline code/__DEFINES/callbacks.dm:12]:
```suggestion
		ASYNC { \
```

File: code/__DEFINES/callbacks.dm
Comment (MrStonedOne) [inline code/__DEFINES/callbacks.dm:17]:
```suggestion
		ASYNC { \
```

File: code/__DEFINES/callbacks.dm
Comment (MrStonedOne) [inline code/__DEFINES/callbacks.dm:11]:
```suggestion
	if ((proc_owner) == GLOBAL_PROC) { \
```

**Review by MrsTonedOne — APPROVED** [review]

**Review by LemonInTheDark — APPROVED** [review]

**Comment by LemonInTheDark** [comment]
testmerging for performance testing

File: code/modules/vehicles/mecha/mecha_construction_paths.dm
Comment (ZephyrTFA) [inline code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: Couldn't this just be
```
		return check_step(used_atom, user)
```

Comment (Kapu1178) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: check_step calls do_after. this triggers the linter due to SHPULD NOT SLEEP

Comment (MrStonedOne) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: > Couldn't this just be

returns inside of spawn blocks is undefined behavior.

in this case i think it would actually work like a break, and move control to the end of the spawn block.

Comment (MrStonedOne) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: actually...

image

Comment (Kapu1178) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: So it's the same rules as proc arguments. Objects are pass-by-reference, everything else is copied. Interesting. Ideas?

Comment (LemonInTheDark) [reply code/modules/vehicles/mecha/mecha_construction_paths.dm:389]: we shouldn't be trying to bypass the linter like this
optimally we'd have a way to mark a proc call as DOES_NOT_SLEEP, but an UNLINT could also work

The offending case here should not ever sleep, at least in the way it currently fails, but we might end up missing other cases.
S annoying.

File: code/__DEFINES/callbacks.dm
Comment (ZeWaka) [inline code/__DEFINES/callbacks.dm:20]: Why are you escaping this newline? Seems unnecessary - if someone eventually puts something below it'll eat it

Comment (Kapu1178) [reply code/__DEFINES/callbacks.dm:20]: error

**Comment by MrStonedOne** [comment]
image

File: code/__DEFINES/callbacks.dm
Comment (MrStonedOne) [inline code/__DEFINES/callbacks.dm:21]:
```suggestion
		}; \
	} 

```

**Pull request merged**
